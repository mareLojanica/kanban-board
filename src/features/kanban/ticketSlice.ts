import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { ticketsApi } from "../../api/tickets.grapql"

interface TicketsState {
	tickets: Array<{
		id: string
		status: string
		title: string
	}>
	isLoading: boolean
	isError: boolean
	searchText: string
}

const initialState: TicketsState = {
	isLoading: false,
	tickets: [],
	isError: false,
	searchText: "",
}

const ticketsSlice = createSlice({
	name: "tickets",
	initialState,
	reducers: {
		setTickets(
			state,
			action: PayloadAction<
				{
					id: string
					status: string
					title: string
				}[]
			>
		) {
			state.tickets = action.payload
		},
		setSearchText(state, action: PayloadAction<string>) {
			state.searchText = action.payload
		},
		deleteSelectedTicket(state, action: PayloadAction<string>) {
			state.tickets = state.tickets.filter(
				(ticket) => ticket.id !== action.payload
			)
		},
		addNewTicket(
			state,
			action: PayloadAction<{
				id: string
				status: string
				title: string
			}>
		) {
			state.tickets = [...state.tickets, action.payload]
		},
		updateTicketById(
			state,
			action: PayloadAction<{ id: string; title: string }>
		) {
			state.tickets = state.tickets.map((ticket) => {
				if (ticket.id === action.payload.id) {
					return { ...ticket, title: action.payload.title }
				}
				return { ...ticket }
			})
		},
	},
	extraReducers: (builder) => {
		builder.addMatcher(
			ticketsApi.endpoints.getTickets.matchFulfilled,
			(state, { payload }) => {
				state.tickets = payload
			}
		)
	},
})

export const {
	setTickets,
	setSearchText,
	deleteSelectedTicket,
	addNewTicket,
	updateTicketById,
} = ticketsSlice.actions
export default ticketsSlice.reducer
