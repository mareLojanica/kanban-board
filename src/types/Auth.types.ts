export interface AuthState {
	accessToken: string | null
	isLoading: boolean
	error: string | null
}
