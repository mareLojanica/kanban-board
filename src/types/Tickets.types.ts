export interface Ticket {
	id: string
	status: string
	title: string
	isNew?: boolean
}

export interface TicketsState {
	tickets: Ticket[]
	isLoading: boolean
	isError: boolean
	searchText: string
}
