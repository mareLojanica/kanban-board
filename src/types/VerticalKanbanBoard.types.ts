export enum TicketStatus {
	TO_DO = "TO_DO",
	IN_PROGRESS = "IN_PROGRESS",
	DONE = "DONE",
}
export interface KanbanCardProps {
	id: string
	title: string
	index: number
	parent: string
	isNew?: boolean
	cardColor: string
}
export interface ColumnConfig {
	alias: TicketStatus
	headerColor: string
	backgroundColor: string
	cardColor: string
	title: string
}

export interface KanbanLaneProps {
	column: ColumnConfig
	items: {
		title: string
		id: string
		isNew?: boolean
	}[]
}
