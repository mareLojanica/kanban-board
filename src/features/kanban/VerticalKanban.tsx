import {
	DndContext,
	DragEndEvent,
	rectIntersection,
	useSensor,
	useSensors,
} from "@dnd-kit/core"
import {
	useGetTicketsQuery,
	useUpdateTicketMutation,
} from "../../api/tickets.grapql"
import Spinner from "../../components/Spinner"
import {
	Box,
	debounce,
	Input,
	InputProps,
	styled,
	useMediaQuery,
	useTheme,
} from "@mui/material"
import { columnsConfig } from "../../config/kanbanConfig"

import { useDispatch, useSelector } from "react-redux"
import { setSearchText, setTickets } from "./ticketSlice"
import { RootState } from "../../app/store"
import { TicketStatus } from "../../types/VerticalKanbanBoard.types"
import { getEnumValueByKey } from "../../utils"
import { selectSearchText } from "./ticketSelectors"
import { useCallback, useState } from "react"
import { CustomPointerSensor } from "../../utils/CustomPointerSensor"
import KanbanLane from "../../components/KanbanLane"

const StyledInput = styled(Input)<InputProps>(({ theme }) => ({
	borderRadius: "8px",
	border: "1px solid #ccc",
	padding: "8px 12px",
	transition: "border-color 0.3s ease, box-shadow 0.3s ease",
	"&:hover": {
		borderColor: theme.palette.primary.main,
	},
	"&.Mui-focused": {
		borderColor: theme.palette.primary.main,
		boxShadow: `0 0 4px ${theme.palette.primary.main}`,
	},
	"&.Mui-error": {
		borderColor: theme.palette.error.main,
		boxShadow: `0 0 4px ${theme.palette.error.main}`,
	},
	"& input": {
		fontSize: "16px",
		color: theme.palette.text.primary,
	},
}))

const VerticalKanban = () => {
	const theme = useTheme()
	const isMobileOrTablet = useMediaQuery(theme.breakpoints.down("md"))
	const pointerSensor = useSensor(CustomPointerSensor, {
		activationConstraint: {
			delay: 100,
			tolerance: 5,
		},
	})

	const sensors = useSensors(pointerSensor)
	const dispatch = useDispatch()
	const data = useSelector((state: RootState) => state.tickets.tickets)
	const searchText = useSelector(selectSearchText)
	const [debouncedSearchText, setDebouncedSearchText] = useState(searchText)
	const { isLoading } = useGetTicketsQuery({
		searchText:
			debouncedSearchText.length > 3 ? debouncedSearchText : undefined,
	})

	const [updateTicket] = useUpdateTicketMutation()

	const handleSearchChange = useCallback(
		debounce((value: string) => {
			setDebouncedSearchText(value)
		}, 300),
		[]
	)

	if (isLoading) {
		return <Spinner />
	}

	const handleDragEnd = async (event: DragEndEvent) => {
		const draggedTo = event.over?.id
		const activeCard = event.active.id
		if (
			draggedTo &&
			event.active.data.current &&
			draggedTo !== event.active.data.current.parent
		) {
			const updatedTickets = data?.map((ticket) => {
				if (ticket.id === activeCard) {
					return { ...ticket, status: String(draggedTo) }
				} else {
					return { ...ticket }
				}
			})

			dispatch(setTickets(updatedTickets ?? []))
			const statusKey = getEnumValueByKey(TicketStatus, String(draggedTo))
			updateTicket({
				id: String(activeCard),
				status: statusKey,
			}).unwrap()
		}
	}

	const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const value = e.target.value
		dispatch(setSearchText(value))
		handleSearchChange(value)
	}

	return (
		<DndContext
			collisionDetection={rectIntersection}
			onDragEnd={handleDragEnd}
			sensors={sensors}
		>
			<Box
				sx={{
					display: "flex",
					flexDirection: "column",
					p: 2,
				}}
			>
				<Box
					sx={{
						width: "100%",
						display: "flex",
						justifyContent: "center",
					}}
				>
					<StyledInput
						placeholder="Search tickets..."
						sx={{
							width: {
								xs: "100%",
								md: "50%",
							},
						}}
						value={searchText}
						onChange={handleInputChange}
					/>
				</Box>
				<Box
					sx={{
						display: "flex",
						flexDirection: "row",
						gap: "16px",
						height: "100%",
						padding: 2,
						overflow: "auto",
					}}
				>
					{columnsConfig.map((column) => (
						<Box
							key={column.alias}
							sx={{
								width: isMobileOrTablet ? "250px" : "33%",
								minWidth: "250px",
							}}
						>
							<KanbanLane
								column={column}
								items={
									data.filter(
										(card) => card.status === column.alias
									) ?? []
								}
							/>
						</Box>
					))}
				</Box>
			</Box>
		</DndContext>
	)
}

export default VerticalKanban
