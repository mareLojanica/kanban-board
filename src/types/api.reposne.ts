import { Ticket } from "./Tickets.types"
import { TicketStatus } from "./VerticalKanbanBoard.types"
import { MutationOptions, QueryOptions } from "@apollo/client"

export interface TicketResponse {
	id: string
	title: string
	status: TicketStatus
	__typename: string
}
export interface GetTicketsResponse {
	getTickets: TicketResponse[]
}

export interface UpdateTicketResponse {
	updateTicket: Omit<TicketResponse, "__typename">
}

export interface DeleteTicketResponse {
	deleteTicket: {
		success: boolean
	}
}

export interface CreateTicketResponse {
	createTicket: Ticket
}

export interface BaseQueryArgs {
	body:
		| MutationOptions<
				MutationOptions,
				Record<string, any>,
				{ headers: { Authorization: string } }
		  >
		| QueryOptions<any, Record<string, any>>
}
