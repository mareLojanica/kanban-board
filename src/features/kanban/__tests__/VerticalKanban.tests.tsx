import { fireEvent, render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"
import VerticalKanban from "../VerticalKanban"
import { setSearchText, setTickets } from "../ticketSlice"
import {
	useGetTicketsQuery,
	useUpdateTicketMutation,
} from "../../../api/tickets.grapql"
import { TicketStatus } from "../../../types/VerticalKanbanBoard.types"
import { act } from "react"

jest.mock("../../../api/tickets.grapql", () => ({
	useUpdateTicketMutation: jest.fn(),
}))

const mockStore = configureStore([])

jest.mock("../../../components/Spinner", () => () => (
	<div data-testid="spinner" />
))

jest.mock("../../../api/tickets.grapql", () => ({
	useGetTicketsQuery: jest.fn(),
	useUpdateTicketMutation: jest.fn(),
}))

jest.mock("../../../components/KanbanLane", () => ({ column, items }: any) => (
	<div data-testid={`kanban-lane-${column.alias}`}>
		{items.map((item: any) => (
			<div key={item.id} data-testid={`kanban-card-${item.id}`} />
		))}
	</div>
))

jest.mock("../../../utils", () => ({
	...jest.requireActual("../../../utils"),
	debounce: (fn: any) => fn,
}))
const mockGetTicketsQuery = jest.fn(() => ({
	isLoading: false,
	data: [
		{ id: "1", title: "Ticket 1", status: "TO_DO" },
		{ id: "2", title: "Ticket 2", status: "IN_PROGRESS" },
		{ id: "3", title: "Ticket 3", status: "DONE" },
	],
}))
jest.mock("../../../utils", () => ({
	...jest.requireActual("../../../utils"),
	debounce: (fn: any) => fn,
}))

const mockUpdateTicket = jest.fn(() => ({
	unwrap: jest.fn().mockResolvedValue({}),
}))

;(useGetTicketsQuery as jest.Mock).mockImplementation(mockGetTicketsQuery)
;(useUpdateTicketMutation as jest.Mock).mockReturnValue([mockUpdateTicket])

describe("VerticalKanban Component", () => {
	let store

	beforeEach(() => {
		jest.clearAllMocks()
		store = mockStore({
			tickets: {
				tickets: [
					{ id: "1", title: "Ticket 1", status: TicketStatus.TO_DO },
					{
						id: "2",
						title: "Ticket 2",
						status: TicketStatus.IN_PROGRESS,
					},
					{ id: "3", title: "Ticket 3", status: TicketStatus.DONE },
				],
				searchText: "",
			},
		})
	})

	const renderComponent = () =>
		render(
			<Provider store={store}>
				<VerticalKanban />
			</Provider>
		)
	it("renders the VerticalKanban correctly", () => {
		const { container } = renderComponent()
		expect(container).toMatchSnapshot()
	})

	it("filters tickets based on search input", async () => {
		renderComponent()

		const searchInput = screen.getByPlaceholderText("Search tickets...")
		const user = userEvent.setup()

		await user.type(searchInput, "Ticket 1")

		const actions = store.getActions()
		const action = actions.reduce(
			(acc, curr: { type: string; payload: string }) => {
				acc.payload += curr.payload
				acc.type = curr.type
				return acc
			},
			{ payload: "", type: "" }
		)

		expect(action).toEqual(setSearchText("Ticket 1"))
	})

	// it("dispatches setTickets on drag-and-drop", () => {
	// 	renderComponent()

	// 	const kanbanCard = screen.getByTestId("kanban-card-1")
	// 	const targetLane = screen.getByTestId("kanban-lane-IN_PROGRESS")
	// 	const mockEvent = {
	// 		active: { id: "1" },
	// 		over: { id: "IN_PROGRESS" },
	// 	}
	// 	fireEvent.touchStart(kanbanCard)
	// 	fireEvent.dragEnd(kanbanCard, mockEvent)
	// 	fireEvent.touchEnd(kanbanCard)
	// 	const actions = store.getActions()
	// 	expect(actions).toContainEqual(
	// 		setTickets([
	// 			{ id: "1", title: "Ticket 1", status: "IN_PROGRESS" },
	// 			{ id: "2", title: "Ticket 2", status: "IN_PROGRESS" },
	// 			{ id: "3", title: "Ticket 3", status: "DONE" },
	// 		])
	// 	)
	// })
})
