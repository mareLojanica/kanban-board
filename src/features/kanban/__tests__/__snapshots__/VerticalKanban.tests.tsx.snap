// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalKanban Component renders the VerticalKanban correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1w1rbvj"
  >
    <div
      class="MuiBox-root css-1ykdma4"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1q9x765-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
          placeholder="Search tickets..."
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-1s4g6lc"
    >
      <div
        class="MuiBox-root css-4wg6d4"
      >
        <div
          data-testid="kanban-lane-TO_DO"
        >
          <div
            data-testid="kanban-card-1"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-4wg6d4"
      >
        <div
          data-testid="kanban-lane-IN_PROGRESS"
        >
          <div
            data-testid="kanban-card-2"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-4wg6d4"
      >
        <div
          data-testid="kanban-lane-DONE"
        >
          <div
            data-testid="kanban-card-3"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;
